### ISSUE TITLE
[Proposal] Chunk-Caching for SillyProblems

### ISSUE THREAD

**Issue by Robotia** [TCPR/KCauldron#11]
TL; DR Some extremely fast operations (i.e. Redstone Repeater timers) cause unnecessary lag by fetching chunks from the hasmap. The server should cache the last few accessed chunks. I'm open to any ideas for how this collection could be organized.

I have managed to reduce the bottleneck of chunk-getting from the hashmap through implementing #10. However, I have found that frequent,  fast access by mods (and sometimes even vanilla features) makes this bottleneck an even bigger problem. See the examples below (or skip ahead):

1. A player sets off a bunch of massive explosions. The entire server lags as a result. The source lies in the fact that setBlock() to air requires the server to fetch the loaded chunk from the hashmap **every time a block is set**.

2. A player creates an extremely fast timer with a few redstone repeaters. Again, the server lags. The source, once again, lies in the fact that setBlock() to redstone powered/unpowered repeater fetches the loaded chunk **every time**.

3. A player sets a forest on fire. The fire spread call to surrounding tiles and setting surrounding blocks on fire *also* fetches the loaded chunk **every time**.

These problems make no sense to me and they have probably been around forever.

I propose that KCauldron implement a sort of chunk-caching feature for the *most accessed* chunks. A GC flag (similar to the one applied in #8) could be used to prevent cached chunks from being inadvertently accessed when unloaded. An array could be implemented, but the cache-size would limit the speed of operations on chunks outside of the cache. Maybe a small hashmap inside methods that caches chunk calls and clears every 50ms?

This may take some time to implement but will definitely be worthwhile.

**Comment by Robotia** [comment]
Starting work on this, going to throw it on my Tekkit Legends server and see if I get any complaints.

**Comment by Bogdan-G** [comment]
The forge was setting in the configuration to cache chunks.

**Comment by Robotia** [comment]
This is different, caching for call to provideChunk under Chunk Server. This will cache the most recently accessed chunks from the long hash map.

**Comment by Robotia** [comment]
@Bogdan-G Tested it out, it's working amazingly well. Message me on Prok.PW GitLab I'll give you test server IP.

**Comment by Robotia** [comment]
@Bogdan-G Gist of diff -wrc. https://gist.github.com/Robotia/6d7f586d5e861b9f2be9

Not sure how to add these changes.

**Comment by Bogdan-G** [comment]
@Robotia Find files that have been changed through the patch files. Example: https://github.com/TCPR/KCauldron/blob/master/patches/net/minecraft/entity/Entity.java.patch

If you change a few changes to the files, you need to find one or create a file X.java.patch

**Comment by Robotia** [comment]
@Bogdan-G So just type changes into the patch files?

**Comment by Bogdan-G** [comment]
@Robotia Yes, quite a simple design.

+ and - at the beginning of the rows is the removal and addition of rows affected. The format is generally as follows: 1) the fact that remove the (continuous), and immediately after that that we add. 2) ukazvaem not changed rows and between them can go single line with the + or -

**Comment by Robotia** [comment]
@Bogdan-G Successful implementation on Tekkit Legends server ([here](http://technicpack.net/modpack/tekkit-legends.735902)) at tl.accretion.us:25565

**Comment by Yive** [comment]
I've made you able to commit changes to this repository, @Robotia so you can just commit your changes once finished.

**Comment by Robotia** [comment]
@Yive awesome. Also, I made a post on TCPR about a temporary fix for tick synch issues and linked them to an unofficial build [here](https://accretion.us/development).

**Comment by Robotia** [comment]
@Bogdan-G Figured it out, there's a command :cry: 
`git difftool eclipse/Clean/src/main/java eclipse/cauldron/src/main/java`

**Comment by Robotia** [comment]
This proposal has been implemented with a 500 chunk cache size. Closing.
